File: frontend/src/TerraForge/lib/biomes.js
// Biome classification → splatmap (RGBA, each channel = a biome weight)
// R = sand/beach, G = grass, B = rock, A = snow
// Water is implicit (height < waterLevel). Exported separately if needed.

export const DEFAULT_BIOME_PARAMS = {
  waterLevel: 0.12,
  sandWidth: 0.05,
  grassMax: 0.55,
  rockMax: 0.82,
  slopeBias: 0.45,
  blendSoftness: 0.04,
};

function smoothstep(edge0, edge1, x) {
  const t = Math.max(0, Math.min(1, (x - edge0) / (edge1 - edge0 || 1e-6)));
  return t * t * (3 - 2 * t);
}

// Compute slope at pixel using central differences
export function computeSlopeMap(height, size) {
  const slope = new Float32Array(size * size);
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const xl = Math.max(0, x - 1);
      const xr = Math.min(size - 1, x + 1);
      const yu = Math.max(0, y - 1);
      const yd = Math.min(size - 1, y + 1);
      const dx = height[y * size + xr] - height[y * size + xl];
      const dy = height[yd * size + x] - height[yu * size + x];
      slope[y * size + x] = Math.min(1, Math.hypot(dx, dy) * size * 0.05);
    }
  }
  return slope;
}

// Returns { splatmap: Uint8ClampedArray (RGBA), waterMask: Uint8Array, slopeMap: Float32Array }
export function computeBiomeSplatmap(height, size, params = DEFAULT_BIOME_PARAMS) {
  const { waterLevel, sandWidth, grassMax, rockMax, slopeBias, blendSoftness } = params;
  const slope = computeSlopeMap(height, size);
  const splat = new Uint8ClampedArray(size * size * 4);
  const waterMask = new Uint8Array(size * size);

  for (let i = 0; i < size * size; i++) {
    const h = height[i];
    const s = slope[i];
    if (h < waterLevel) {
      waterMask[i] = 255;
      splat[i * 4] = 0;
      splat[i * 4 + 1] = 0;
      splat[i * 4 + 2] = 0;
      splat[i * 4 + 3] = 0;
      continue;
    }
    const hN = (h - waterLevel) / (1 - waterLevel + 1e-6);

    // Weight curves with smoothstep transitions
    const wSand = (1 - smoothstep(sandWidth - blendSoftness, sandWidth + blendSoftness, hN)) * (1 - s * slopeBias);
    const wGrass = smoothstep(sandWidth - blendSoftness, sandWidth + blendSoftness, hN)
                 * (1 - smoothstep(grassMax - blendSoftness, grassMax + blendSoftness, hN))
                 * (1 - s * slopeBias);
    const wRock = smoothstep(grassMax - blendSoftness, grassMax + blendSoftness, hN)
                * (1 - smoothstep(rockMax - blendSoftness, rockMax + blendSoftness, hN))
                + s * slopeBias * 0.7;
    const wSnow = smoothstep(rockMax - blendSoftness, rockMax + blendSoftness, hN);

    const total = wSand + wGrass + wRock + wSnow + 1e-6;
    splat[i * 4] = Math.round((wSand / total) * 255);
    splat[i * 4 + 1] = Math.round((wGrass / total) * 255);
    splat[i * 4 + 2] = Math.round((wRock / total) * 255);
    splat[i * 4 + 3] = Math.round((wSnow / total) * 255);
  }
  return { splatmap: splat, waterMask, slopeMap: slope };
}

// Color preview from splat + height (for the 3D viewer)
export function biomeColorAt(h, s, params = DEFAULT_BIOME_PARAMS) {
  const { waterLevel, sandWidth, grassMax, rockMax, slopeBias, blendSoftness } = params;
  if (h < waterLevel) {
    const d = (waterLevel - h) / (waterLevel + 1e-6);
    return [0.05 + 0.05 * (1 - d), 0.18 + 0.10 * (1 - d), 0.28 + 0.10 * (1 - d)];
  }
  const hN = (h - waterLevel) / (1 - waterLevel + 1e-6);
  const wSand = (1 - smoothstep(sandWidth - blendSoftness, sandWidth + blendSoftness, hN)) * (1 - s * slopeBias);
  const wGrass = smoothstep(sandWidth - blendSoftness, sandWidth + blendSoftness, hN)
               * (1 - smoothstep(grassMax - blendSoftness, grassMax + blendSoftness, hN))
               * (1 - s * slopeBias);
  const wRock = smoothstep(grassMax - blendSoftness, grassMax + blendSoftness, hN)
              * (1 - smoothstep(rockMax - blendSoftness, rockMax + blendSoftness, hN))
              + s * slopeBias * 0.7;
  const wSnow = smoothstep(rockMax - blendSoftness, rockMax + blendSoftness, hN);
  const total = wSand + wGrass + wRock + wSnow + 1e-6;
  const cSand = [0.78, 0.68, 0.45];
  const cGrass = [0.32, 0.46, 0.22];
  const cRock = [0.42, 0.38, 0.34];
  const cSnow = [0.95, 0.95, 0.97];
  return [
    (cSand[0] * wSand + cGrass[0] * wGrass + cRock[0] * wRock + cSnow[0] * wSnow) / total,
    (cSand[1] * wSand + cGrass[1] * wGrass + cRock[1] * wRock + cSnow[1] * wSnow) / total,
    (cSand[2] * wSand + cGrass[2] * wGrass + cRock[2] * wRock + cSnow[2] * wSnow) / total,
  ];
}

File: TerraForgeRust/core/src/biomes.rs
//! Biome classification → RGBA splatmap. Port of `lib/biomes.js`.

use crate::smoothstep;

#[derive(Clone, Debug)]
pub struct BiomeParams {
    pub water_level: f32,
    pub sand_width: f32,
    pub grass_max: f32,
    pub rock_max: f32,
    pub slope_bias: f32,
    pub blend_softness: f32,
}

impl Default for BiomeParams {
    fn default() -> Self {
        Self {
            water_level: 0.12,
            sand_width: 0.05,
            grass_max: 0.55,
            rock_max: 0.82,
            slope_bias: 0.45,
            blend_softness: 0.04,
        }
    }
}

pub fn compute_slope_map(h: &[f32], size: u32) -> Vec<f32> {
    let n = size as usize;
    let mut slope = vec![0f32; n * n];
    for y in 0..n {
        for x in 0..n {
            let xl = x.saturating_sub(1);
            let xr = (x + 1).min(n - 1);
            let yu = y.saturating_sub(1);
            let yd = (y + 1).min(n - 1);
            let dx = h[y * n + xr] - h[y * n + xl];
            let dy = h[yd * n + x] - h[yu * n + x];
            slope[y * n + x] = ((dx * dx + dy * dy).sqrt() * n as f32 * 0.05).min(1.0);
        }
    }
    slope
}

pub struct BiomeResult {
    pub splatmap: Vec<u8>,    // RGBA
    pub water_mask: Vec<u8>,  // 0 / 255
    pub slope_map: Vec<f32>,
}

/// Compute the RGBA splatmap (R=sand, G=grass, B=rock, A=snow) + water mask + slope.
pub fn compute_biome_splatmap(h: &[f32], size: u32, p: &BiomeParams) -> BiomeResult {
    let n = (size * size) as usize;
    let slope = compute_slope_map(h, size);
    let mut splat = vec![0u8; n * 4];
    let mut water = vec![0u8; n];

    for i in 0..n {
        let height = h[i];
        let s = slope[i];
        if height < p.water_level {
            water[i] = 255;
            continue; // splat already zeros
        }
        let hn = (height - p.water_level) / (1.0 - p.water_level + 1e-6);
        let w_sand = (1.0 - smoothstep(p.sand_width - p.blend_softness, p.sand_width + p.blend_softness, hn))
            * (1.0 - s * p.slope_bias);
        let w_grass = smoothstep(p.sand_width - p.blend_softness, p.sand_width + p.blend_softness, hn)
            * (1.0 - smoothstep(p.grass_max - p.blend_softness, p.grass_max + p.blend_softness, hn))
            * (1.0 - s * p.slope_bias);
        let w_rock = smoothstep(p.grass_max - p.blend_softness, p.grass_max + p.blend_softness, hn)
            * (1.0 - smoothstep(p.rock_max - p.blend_softness, p.rock_max + p.blend_softness, hn))
            + s * p.slope_bias * 0.7;
        let w_snow = smoothstep(p.rock_max - p.blend_softness, p.rock_max + p.blend_softness, hn);
        let total = w_sand + w_grass + w_rock + w_snow + 1e-6;
        splat[i * 4] = ((w_sand / total) * 255.0).round() as u8;
        splat[i * 4 + 1] = ((w_grass / total) * 255.0).round() as u8;
        splat[i * 4 + 2] = ((w_rock / total) * 255.0).round() as u8;
        splat[i * 4 + 3] = ((w_snow / total) * 255.0).round() as u8;
    }
    BiomeResult { splatmap: splat, water_mask: water, slope_map: slope }
}
